### ISSUE TITLE
Generalize stats reporting so that backends other than DataDog can be used, if desired

### ISSUE THREAD

**Issue by wyhaines** [forem/forem#12005]
Currently, Forem's statistics collection is tied directly to DataDog. I would like to use New Relic instead of DataDog without requiring any customizations to the deployed code, so that a managed Forem service can support both DataDog and NewRelic (and, eventually, other providers) for stats collection.

The `DatadogStatsClient` will be turned into a `ForemStatsClient`. It is the high level interface for Forem to record metrics and events. The default mode of operation will be for the `ForemStatsClient` to use a `DatadogStatsDriver` to invoke the exact same API calls that are invoked currently.

A conditional section will be added to the `Gemfile`. If a New Relic configuration is present, it will include the New Relic gem, but if it is not, it will include the Datadog gem. The `ForemStatsClient` likewise does a conditional require for loading it's driver. Thus, if there is no New Relic configuration, everything will still use DataDog unchanged, just as it does currently. This work will be transparent to all current Forem users.

If there is a New Relic config, though, then Forem will use New Relic for recording metrics and events. The sticky wicket to the plan is really just that while the New Relic platform supports all of the things that the Forem software wants to do with recording of events and metrics, the New Relic Ruby Gem does not. A subset of things cleanly translate. The rest are a dirty translation -- the limitations of the Ruby client mean that we either send the data to New Relic in a way that may be suboptimal for our own experience interacting with the data in the New Relic Dashboard, or we go the hard route and implement code to directly use the HTTP API for New Relic to do the things that the Ruby client won't do, or we lose data. My first pass is to do cleanly what we can do cleanly, and make a best effort at everything else, with NOOP stubs for the things that are hard. The goal is to get something that can be tested and validated and that provides some value quickly, which can then be iterated on to make it better.

The actual `NewRelicStatsDriver` is to be implemented as a separate gem, instead of being bundled into Forem. That way we can make incremental improvements to the integration with New Relic without making code PRs to Forem itself, and both the maintenance and support for the New Relic integration can be cleanly separated from the maintenance and support of the Forem product itself, helping to insulate Forem from increased support burden as a result of this change.

This is being implemented as a couple of discrete PRs. The first creates the `ForemStatsClient` and the `DatadogStatsDriver`.

The second PR adds the conditional section to the `Gemfile` for enabling the use of New Relic in the presence of a New Relic configuration.

In parallel to this, a well tested gem to implement the `NewRelicStatsDriver` is being prepared.

The end result should be that all existing Forem users will see no operational difference, but if a New Relic configuration is added to a Forem deployment, it will use New Relic for stats collection instead of DataDog. The end result should also be generic enough that other stats vendors can write drivers for their service, as well.

I am implementing this feature myself, and will provide PRs.

**Comment by wyhaines** [comment]
To be clear, I started this implementation a couple weeks ago. I am planning on giving you a PR for the first half of the implementation this week (today, maybe), that implements all of the groundwork for the stats reporting abstractions, but without changing any of the final behavior. It turns the `DatadogStatsClient` into a `ForemStatsClient`, and adds a `DatadogStatsDriver`.

Once that piece is acceptable for merging, I'll follow it up with the changes to enable the New Relic interop as a separate PR.

**Comment by wyhaines** [comment]
Quick update. Over the holiday break I ended up taking an almost total break from the keyboard. I'm prepping this PR for delivery here today.

**Comment by wyhaines** [comment]
We had a wee bit of a delay, both with my own scheduling and then with getting everything running locally. The PR is available for review. Please let me know if there are any questions or concerns.

**Comment by wyhaines** [comment]
All of the review changes were addressed, and specs are happy:

https://github.com/forem/forem/pull/12304

**Comment by cmgorton** [comment]
Just seeing this issue. @wyhaines would it be fair to say the PR you added and was merged should close this issue? Are there still things that need to be changed before we can close it? Thanks for your input🙏
